### ISSUE TITLE
Custom migration source is always overwritten with FsMigrations

### ISSUE THREAD

**Issue by KrishnaPG** [knex/knex#3834]
# Environment

Knex version: 0.21.1
Database + version: SQLite 3
OS: Windows 10

# Bug

When using a customer migration source, the call to `make` method is not using the supplied custom source. Rather it always goes back to the built-in `FsMigrations`.

After debugging, the culprit seems to be at this line: https://github.com/knex/knex/blob/fe8340a6274202b5c59cb623f3e387aa7801f569/lib/migrate/Migrator.js#L298 

The correct code should be:
````
  make(name, config) {
    return this.generator.make(name, config);  //<- should send config, not this.config
  }
````
### Background
To give a background, I am creating a memory-backed migration source that does *not* use file-system. But unable to make the Knex use it, since the `FsMigrations` is always kicking-in.

That is because down the line in the series of calls, a call to `getMergedConfig(this.config, this.config)` ends up clearing the user-supplied custom migration source and resetting it to the built-in `FsMigrations`. Here is that part of the code, just for reference: 
https://github.com/knex/knex/blob/fe8340a6274202b5c59cb623f3e387aa7801f569/lib/migrate/configuration-merger.js#L27-L42

**Comment by elhigu** [comment]
Please add reproduction code. IIRC there is a test for non-FsMigrations based migration source.

**Comment by KrishnaPG** [comment]
Please try the below to reproduce the bug:
````

class MigrationSource {
	getMigrations() {
		console.log("getMigrations got called: ");
		return Promise.resolve(['migration1'])
	}
	getMigrationName(migration) {
		console.log("getMigrationName: ", migration);
		return migration;
	}
	getMigration(migration) {
		console.log("getMigration got called with: ", migration);
	}
};

const migrationSource = new MigrationSource();

const knex = require('knex')({
	client: 'sqlite3',
	connection: { filename: './.db/db.sqlite' },
	migrations: { migrationSource }
});

knex.migrate.make("test1", {  });
````
If it is working, then you should see console log output.

**Comment by KrishnaPG** [comment]
Also, after struggling for the past few days working on this custom migration source implementation, it turns out that this whole API section needs complete refactoring. 

For example, the `getMigrations()` method supports promises, but the `getMigrationName()` and `getMigration()` does not support promises, nor even callbacks (and rather they expect the result to be returned from these API calls in synchronous fashion) !!

**Comment by kibertoad** [comment]
@KrishnaPG I finally have some time, will take a look today.

**Comment by kibertoad** [comment]
@KrishnaPG There might be another issue with config override, I will look into it after merging first refactoring PR. Tests might have missed that, since they call migrator directly, bypassing the CLI part.

**Comment by kibertoad** [comment]
@KrishnaPG See #3839
Can you double-check that you don't have FS-related configs in your knexfile that would result in your migration source being reset?

**Comment by KrishnaPG** [comment]
Thank you @kibertoad 

I do not have any configs at all except for indicating the client connection and migration source. 

The reason why this problem happens is when you call `knex.migrate.make()` the code executes as below: https://github.com/knex/knex/blob/fe8340a6274202b5c59cb623f3e387aa7801f569/lib/migrate/Migrator.js#L296-L300

Note the code `this.generator.make(name, this.config)`. Now notice what happens inside the `generator.make` code https://github.com/knex/knex/blob/fe8340a6274202b5c59cb623f3e387aa7801f569/lib/migrate/MigrationGenerator.js#L12-L14

- The code is executing `getMergedConfig` in sequence immediately two times
- As a result, essentially you are calling `getMergedConfig(this.config, this.config)` i.e. merging `this.config` with itself.

This triggers the configuration-merger.js to reset the custom source https://github.com/knex/knex/blob/fe8340a6274202b5c59cb623f3e387aa7801f569/lib/migrate/configuration-merger.js#L35
which is then defaulted back to `fsMigrations` https://github.com/knex/knex/blob/fe8340a6274202b5c59cb623f3e387aa7801f569/lib/migrate/configuration-merger.js#L41

------------------
Fix:
---------------------
The **fix** for this is: **remove the Line 297** in the `Migrator.js` file and use the code as below. It solves the problem.
````
  make(name, config) {
    return this.generator.make(name, config);  //<- should send config, not this.config
  }
````

Now, if you are wondering what is wrong with merging `this.config` with itself (apart from it being redundant), please see this default initialization in the config merger file:

https://github.com/knex/knex/blob/fe8340a6274202b5c59cb623f3e387aa7801f569/lib/migrate/configuration-merger.js#L6-L15

Note how a `directory` is being specified by default. This one will kick-in when the below comparison is made to reset the custom migration source. https://github.com/knex/knex/blob/fe8340a6274202b5c59cb623f3e387aa7801f569/lib/migrate/configuration-merger.js#L29-L32

In other words, I never had to specify any file source. The default configuration inside the knex code is forcing the FS.

**Comment by sethlivingston** [comment]
I'm having this problem. Setting

```json
  "migrations": {
    "directory": "pathABC"
  }
```

And then running the CLI with

```
knex migrate:make --migrations-directory pathXYZ test
```

Puts the new migration in pathABC. I would expect the CLI argument to override the knexfile.

When I looked through the code, it doesn't look like the --migrations-directory option is being used anywhere.

**Comment by elhigu** [comment]
I wonder if that directory from command line is just added to the list of migration directories. IIRC there can be multiple migration directories and new migrations are added to the first listed directory.

**Comment by hallais** [comment]
This fix still in progress?

**Comment by kibertoad** [comment]
Appreciate the ping. I'll take a look today or tomorrow.

**Comment by kibertoad** [comment]
Fix released in 0.21.6
